File: ocaml/words/test/test_words.ml
open OUnit2
module Words = Liby.Words

let suite_words = "Words Suite" >:::
  [
    "reverse string" >:: (fun _ ->
      assert_equal "olleh" (Words.reverse_string "hello");
      assert_equal "a" (Words.reverse_string "a");
      assert_equal "" (Words.reverse_string "");
    );

    "is palindrome" >:: (fun _ ->
      assert_equal true (Words.is_palindrome "race car");
      assert_equal true (Words.is_palindrome "level");
      assert_equal false (Words.is_palindrome "hello");
      assert_equal true (Words.is_palindrome "a");
      assert_equal true (Words.is_palindrome "");
    );

    "is substring" >:: (fun _ ->
      assert_equal true (Words.kmp_search "test" "this is a test");
      assert_equal false (Words.kmp_search "not" "this is a test");
      assert_equal true (Words.kmp_search "" "any string");
      assert_equal true (Words.kmp_search "abc" "abc");
      assert_equal false (Words.kmp_search "abc" "ab");
    );

    "longest common subsequence" >:: (fun _ ->
      assert_equal 4 (Words.lcs "AGGTAB" "GXTXAYB");
      assert_equal 3 (Words.lcs "ABC" "ABC");
      assert_equal 0 (Words.lcs "ABC" "DEF");
      assert_equal 0 (Words.lcs "" "ABC");
      assert_equal 0 (Words.lcs "ABC" "");
    );
  ]

let () =
  run_test_tt_main suite_words

File: _experimental/_nim/words/tests/test_words.nim
import unittest
import words

suite "Words Suite":
    test "reverse string":
        assert reverseString("hello") == "olleh"
        assert reverseString("a") == "a"
        assert reverseString("") == ""
    
    test "is palindrome":
        assert isPalindrome("race car") == true
        assert isPalindrome("level") == true
        assert isPalindrome("hello") == false
        assert isPalindrome("a") == true
        assert isPalindrome("") == true
    
    test "is substring":
        assert kmpSearch("test", "this is a test") == true
        assert kmpSearch("not", "this is a test") == false
        assert kmpSearch("", "any string") == true
        assert kmpSearch("abc", "abc") == true
        assert kmpSearch("abc", "ab") == false

    test "longest common subsequence":
        assert lcs("AGGBTAB", "GXTXAYB") == 4
        assert lcs("ABC", "ABC") == 3
        assert lcs("ABC", "DEF") == 0
        assert lcs("", "ABC") == 0
        assert lcs("ABC", "") == 0